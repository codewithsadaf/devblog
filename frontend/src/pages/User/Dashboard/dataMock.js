export const postsMock = [
    {
        _id: "1",
        image: "example.png",
        title: "post title 1",
        content: "<p>post content</p>",
        user: {
            name: "john doe",
        },
        tags: ["#tag1", "#tag2"],
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
    {
        _id: "2",
        image: "example.png",
        title: "post title 2",
        content: "<p>post content</p>",
        user: {
            name: "john doe",
        },
        tags: ["#tag3", "#tag4"],
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
    {
        _id: "3",
        image: "example.png",
        title: "post title 3",
        content: "<p>post content</p>",
        user: {
            name: "john doe",
        },
        tags: ["#tag1", "#tag3"],
        createdAt: "2023-04-05T19:27:03.125+00:00",
        updatedAt: "2023-04-05T19:27:03.125+00:00",
    },
];

export const commentsMock = [
    {
        user: {
            image: "example.png",
            name: "john doe",
        },
        post: {
            _id: "1",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        _id: "1",
        commentText: "this is a comment 01",
    },
    {
        user: {
            image: "example.png",
            name: "john doe",
            _id: "2",
        },
        post: {
            _id: "2",
        },
        createdAt: "2023-04-05T19:27:03.125+00:00",
        _id: "2",
        commentText: "this is a comment 02",
    },
];

export const currentUserMock = {
    currentUser: {
        name: "john doe",
        image: "image.png",
        email: "john@test.com",
        createdAt: "2023-04-05T19:27:03.125+00:00",
    },
};
