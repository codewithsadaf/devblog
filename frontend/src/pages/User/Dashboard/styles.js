//imports
import styled, { css } from "styled-components";
//components styles
import { ImageElm } from "../../../components/Layout/ImageBox/styles";
import { Container as Heading } from "../../../components/Layout/Heading/styles";
import { ButtonContainer as Button } from "../../../components/Layout/Button/styles";
import { InputComponent } from "../../../components/Form/Input/styles";
//react icons
import { FaCog } from "react-icons/fa";
import { Link } from "react-router-dom";

export const Container = styled.div`
    ${({ theme }) => css`
        width: 80%;
        margin: 3em auto 0;

        @media ${theme.media.lteMedium} {
            width: 100%;
        }
    `}
`;

export const TitleWrapper = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    margin-bottom: 1.2em;
`;

export const ProfileWrapper = styled.div`
    ${({ theme }) => css`
        display: flex;
        align-items: center;
        justify-content: space-between;

        @media ${theme.media.lteMedium} {
            flex-direction: column;
        }
    `}
`;

export const ProfileImageWrapper = styled.div`
    ${({ theme }) => css`
        flex-basis: 40%;

        ${ImageElm} {
            width: 300px;
            height: 300px;
            object-fit: fill;
            border-radius: 50%;

            @media ${theme.media.lteMedium} {
                width: 150px;
                height: 150px;
            }
        }

        @media ${theme.media.lteMedium} {
            flex-basis: 100%;
        }
    `}
`;

export const ProfileInformationWrapper = styled.div`
    ${({ theme }) => css`
        flex-basis: 58%;
        min-height: 250px;
        text-align: left;

        ${Heading}:nth-of-type(2) {
            margin: 0.3em 0;
        }

        @media ${theme.media.lteMedium} {
            flex-basis: 100%;
            width: 100%;
            margin-left: 0;
        }
    `}
`;

export const RowStart = styled.div`
    display: flex;
    align-items: center;
    justify-content: flex-start;
    flex-wrap: wrap;
`;

export const CogLink = styled(Link)`
    ${({ theme }) => css`
        font-size: 1.2em;
        margin-left: auto;
        cursor: pointer;
        color: ${theme.colors.primaryColor};
        opacity: 1;
        transition: ${theme.transition.tran_05};

        &:hover {
            opacity: 0.6;
            transition: ${theme.transition.tran_05};
        }
    `}
`;

export const CogIcon = styled(FaCog)`
    ${({ theme }) => css``}
`;

export const PostsCommentsInfoWrapper = styled.div`
    ${({ theme }) => css`
        display: flex;
        width: 100%;
        border: 1px solid ${theme.colors.gray};
        border-radius: 7px;
        padding: 20px;
        margin-top: 1em;
    `}
`;

export const PostsInfoWrapper = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-right: 2em;
`;

export const CommentsInfoWrapper = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
`;

export const ViewOptionWrapper = styled.div`
    ${({ theme }) => css`
        margin: 2em 0;
        display: flex;
        border-bottom: 2px solid ${theme.colors.primaryColor};
    `}
`;

export const ViewOption = styled.span`
    ${({ theme }) => css`
        background-color: transparent;
        color: ${theme.colors.primaryColor};
        padding: 7px;
        font-size: ${theme.font.sizes.medium};
        border-radius: 7px 7px 0 0;
        font-weight: 500;
        cursor: pointer;
        transition: ${theme.transition.tran_03};

        &.active {
            background-color: ${theme.colors.primaryColor};
            color: ${theme.colors.white};
            transition: ${theme.transition.tran_03};
            pointer-events: none;
        }
    `}
`;

export const DashboardHeader = styled.div`
    ${({ theme }) => css`
        display: flex;
        justify-content: space-between;
        align-items: center;
        width: 100%;
        border-bottom: 2px solid ${theme.colors.gray};
        font-weight: 700;
        padding: 10px;
    `}
`;

export const DashboardText = styled.span`
    ${({ theme }) => css`
        font-size: ${theme.font.sizes.default};
        line-height: ${theme.font.line_height.default};
    `}
`;

export const DashboardRow = styled.div`
    ${({ theme }) => css`
        display: flex;
        justify-content: space-between;
        align-items: center;
        padding: 10px;
        width: 100%;
        border-bottom: 2px solid ${theme.colors.lightGray};

        & ${DashboardText} {
            display: block;
            width: 60%;
            word-wrap: break-word;

            @media ${theme.media.lteMedium} {
                width: 100%;
            }
        }

        @media ${theme.media.lteMedium} {
            flex-direction: column;
            justify-content: flex-start;
            align-items: flex-start;
        }
    `}
`;

export const DashboardActionsWrapper = styled.div`
    ${({ theme }) => css`
        margin-left: 2em;
        display: flex;
        justify-content: space-between;
        align-items: center;

        & ${Button} {
            margin-left: 0.5em;

            @media ${theme.media.lteMedium} {
                margin-left: 0;
                margin-right: 0.5em;
            }
        }

        @media ${theme.media.lteMedium} {
            margin-left: 0;
            margin: 1em 0;
        }
    `}
`;

export const CenterTextWrapper = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    text-align: center;
    padding: 2em 0;
`;

export const EditInput = styled(InputComponent)`
    border: none;
    width: 100%;

    &:focus {
        border: none;
    }
`;
