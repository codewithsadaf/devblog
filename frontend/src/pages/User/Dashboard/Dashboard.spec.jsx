//imports
import { fireEvent, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
//styles
import { renderTheme } from "../../../styles/render-theme";
//components
import { Dashboard } from ".";
//data mocks
import { commentsMock, postsMock, currentUserMock } from "./dataMock";
//contexts
import { Context as UserContext } from "../../../context/UserContext";
import { PostContext } from "../../../context/PostContext";
import { CommentContext } from "../../../context/CommentContext";

// mock useNavigate
let mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUsedNavigate,
}));

//mock scrollTo
Object.defineProperty(window, "scrollTo", { value: jest.fn() });

describe("<Dashboard/>", () => {
    let userContextValues = {};
    let postContextValues = {};
    let commentContextValues = {};

    beforeEach(() => {
        //contexts values mock
        userContextValues = {
            authenticated: true,
            checkUser: jest.fn().mockReturnValue(
                Promise.resolve({
                    ...currentUserMock,
                })
            ),
        };

        postContextValues = {
            removePost: jest.fn(),
            getUserPosts: jest
                .fn()
                .mockReturnValue(Promise.resolve([...postsMock])),
        };

        commentContextValues = {
            getUserComments: jest
                .fn()
                .mockReturnValue(Promise.resolve([...commentsMock])),
            removeComment: jest.fn(),
            editComment: jest.fn(),
        };

        //clear all "mockedUsedNavigate" calls
        mockedUsedNavigate.mockClear();
    });

    describe("Rendering", () => {
        it("should render Dashboard Component", async () => {
            renderTheme(
                <BrowserRouter>
                    <UserContext.Provider value={{ ...userContextValues }}>
                        <PostContext.Provider value={{ ...postContextValues }}>
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            await waitFor(() => {
                expect(userContextValues.checkUser).toHaveBeenCalledTimes(1);
                expect(postContextValues.getUserPosts).toHaveBeenCalledTimes(1);
                expect(
                    commentContextValues.getUserComments
                ).toHaveBeenCalledTimes(1);
            });

            expect(
                screen.getByRole("img", {
                    name: currentUserMock.currentUser.name,
                })
            ).toHaveAttribute("src", currentUserMock.currentUser.image);
            expect(
                screen.getByRole("heading", {
                    name: currentUserMock.currentUser.name,
                })
            ).toBeInTheDocument();
            expect(screen.getByLabelText("Cog Icon")).toBeInTheDocument();
            expect(
                screen.getByRole("heading", {
                    name: currentUserMock.currentUser.email,
                })
            ).toBeInTheDocument();
            expect(
                screen.getByRole("heading", {
                    name: currentUserMock.currentUser.email,
                })
            ).toBeInTheDocument();
            expect(
                screen.getByRole("heading", {
                    name: `Data de entrada: 5 de Abril 2023`,
                })
            ).toBeInTheDocument();
            expect(screen.getAllByLabelText("View Option")).toHaveLength(2);
        });

        it("should not render Dashboard Component if authenticated is false", () => {
            renderTheme(
                <BrowserRouter>
                    <UserContext.Provider
                        value={{ ...userContextValues, authenticated: false }}
                    >
                        <PostContext.Provider value={{ ...postContextValues }}>
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            expect(
                screen.queryByRole("img", {
                    name: currentUserMock.currentUser.name,
                })
            ).not.toBeInTheDocument();
            expect(
                screen.queryByRole("heading", {
                    name: currentUserMock.currentUser.name,
                })
            ).not.toBeInTheDocument();
            expect(screen.queryByLabelText("Cog Icon")).not.toBeInTheDocument();
            expect(
                screen.queryByRole("heading", {
                    name: currentUserMock.currentUser.email,
                })
            ).not.toBeInTheDocument();
            expect(
                screen.queryByRole("heading", {
                    name: currentUserMock.currentUser.email,
                })
            ).not.toBeInTheDocument();
            expect(
                screen.queryByRole("heading", {
                    name: `Data de entrada: 5 de Abril 2023`,
                })
            ).not.toBeInTheDocument();
            expect(screen.queryAllByRole("View Option")).not.toHaveLength(2);
        });

        it("should render no posts message when getUserPosts return an empty array", async () => {
            postContextValues.getUserPosts = jest
                .fn()
                .mockReturnValue(Promise.resolve([]));
            renderTheme(
                <BrowserRouter>
                    <UserContext.Provider value={{ ...userContextValues }}>
                        <PostContext.Provider
                            value={{
                                ...postContextValues,
                            }}
                        >
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            await waitFor(() => {
                expect(userContextValues.checkUser).toHaveBeenCalledTimes(1);
                expect(postContextValues.getUserPosts).toHaveBeenCalledTimes(1);
                expect(
                    commentContextValues.getUserComments
                ).toHaveBeenCalledTimes(1);
            });

            expect(
                screen.getByRole("heading", {
                    name: "Você ainda não tem nenhum post.",
                })
            ).toBeInTheDocument();
        });
        it("should render no comments message when getUserComments return an empty array", async () => {
            commentContextValues.getUserComments = jest
                .fn()
                .mockReturnValue(Promise.resolve([]));
            renderTheme(
                <BrowserRouter>
                    <UserContext.Provider value={{ ...userContextValues }}>
                        <PostContext.Provider
                            value={{
                                ...postContextValues,
                            }}
                        >
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            await waitFor(() => {
                expect(userContextValues.checkUser).toHaveBeenCalledTimes(1);
                expect(postContextValues.getUserPosts).toHaveBeenCalledTimes(1);
                expect(
                    commentContextValues.getUserComments
                ).toHaveBeenCalledTimes(1);
            });

            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            expect(
                screen.getByRole("heading", {
                    name: "Você ainda não tem nenhum comentário.",
                })
            ).toBeInTheDocument();
        });

        it("should match snapshot", async () => {
            const { container } = renderTheme(
                <BrowserRouter>
                    <UserContext.Provider value={{ ...userContextValues }}>
                        <PostContext.Provider value={{ ...postContextValues }}>
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            await waitFor(() => {
                expect(userContextValues.checkUser).toHaveBeenCalledTimes(1);
                expect(postContextValues.getUserPosts).toHaveBeenCalledTimes(1);
                expect(
                    commentContextValues.getUserComments
                ).toHaveBeenCalledTimes(1);
            });

            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe("Functionality", () => {
        beforeEach(async () => {
            renderTheme(
                <BrowserRouter>
                    <UserContext.Provider value={{ ...userContextValues }}>
                        <PostContext.Provider value={{ ...postContextValues }}>
                            <CommentContext.Provider
                                value={{ ...commentContextValues }}
                            >
                                <Dashboard />
                            </CommentContext.Provider>
                        </PostContext.Provider>
                    </UserContext.Provider>
                </BrowserRouter>
            );

            await waitFor(() => {
                expect(userContextValues.checkUser).toHaveBeenCalledTimes(1);
                expect(postContextValues.getUserPosts).toHaveBeenCalledTimes(1);
                expect(
                    commentContextValues.getUserComments
                ).toHaveBeenCalledTimes(1);
            });
        });

        it("should change view option on button click", async () => {
            //Posts view option active
            expect(screen.getAllByLabelText("View Option")[0]).toHaveClass(
                "active"
            );
            expect(screen.getAllByText(/post title/)).toHaveLength(3);
            expect(screen.getAllByLabelText("View Option")[1]).not.toHaveClass(
                "active"
            );

            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            //Comments view option active
            expect(screen.getAllByLabelText("View Option")[0]).not.toHaveClass(
                "active"
            );
            expect(screen.getAllByLabelText("View Option")[1]).toHaveClass(
                "active"
            );
            expect(screen.getAllByText(/this is a comment/)).toHaveLength(2);
        });

        it("should call redirect function on read post button click", async () => {
            //posts on screen
            expect(screen.getAllByText(/post title/)).toHaveLength(3);

            //click on the first row read post button
            act(() => {
                fireEvent.click(
                    screen
                        .getByText(postsMock[0].title)
                        .parentElement.querySelectorAll("button")[0]
                );
            });

            await waitFor(() => {
                expect(mockedUsedNavigate).toHaveBeenCalledWith(
                    `/post/${postsMock[0]._id}`
                );
            });
        });
        it("should call redirect function on edit post button click", async () => {
            //posts on screen
            expect(screen.getAllByText(/post title/)).toHaveLength(3);

            //click on the first row read post button
            act(() => {
                fireEvent.click(
                    screen
                        .getByText(postsMock[0].title)
                        .parentElement.querySelectorAll("button")[1]
                );
            });
            await waitFor(() => {
                expect(mockedUsedNavigate).toHaveBeenCalledWith(
                    `/post/edit/${postsMock[0]._id}`
                );
            });
        });
        it("should call removePost function on remove post button click", async () => {
            //posts on screen
            expect(screen.getAllByText(/post title/)).toHaveLength(3);

            //click on the first row read post button
            act(() => {
                fireEvent.click(
                    screen
                        .getByText(postsMock[0].title)
                        .parentElement.querySelectorAll("button")[2]
                );
            });
            await waitFor(() => {
                expect(postContextValues.removePost).toHaveBeenCalledTimes(1);
            });
        });

        it("should redirect to post page on see comment button click", async () => {
            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            act(() => {
                fireEvent.click(
                    screen
                        .getByText(commentsMock[0].commentText)
                        .parentElement.querySelectorAll("button")[0]
                );
            });

            await waitFor(() => {
                expect(mockedUsedNavigate).toHaveBeenCalledWith(
                    `/post/${commentsMock[0].post._id}`
                );
            });
        });

        it("should display save button and cancel button on edit comment button click", async () => {
            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            act(() => {
                fireEvent.click(
                    screen
                        .getByText(commentsMock[0].commentText)
                        .parentElement.querySelectorAll("button")[1]
                );
            });

            expect(
                screen.getByRole("button", { name: "Salvar" })
            ).toBeInTheDocument();
            expect(
                screen.getByRole("button", { name: "Cancelar" })
            ).toBeInTheDocument();
        });

        it("should hide save button and cancel button on cancel comment edit button click", async () => {
            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            act(() => {
                fireEvent.click(
                    screen
                        .getByText(commentsMock[0].commentText)
                        .parentElement.querySelectorAll("button")[1]
                );
            });

            expect(
                screen.getByRole("button", { name: "Salvar" })
            ).toBeInTheDocument();
            expect(
                screen.getByRole("button", { name: "Cancelar" })
            ).toBeInTheDocument();

            act(() => {
                fireEvent.click(
                    screen.getByRole("button", { name: "Cancelar" })
                );
            });

            expect(
                screen.queryByRole("button", { name: "Salvar" })
            ).not.toBeInTheDocument();
            expect(
                screen.queryByRole("button", { name: "Cancelar" })
            ).not.toBeInTheDocument();
        });

        it("should call editComment on save comment edit button click", async () => {
            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            act(() => {
                fireEvent.click(
                    screen
                        .getByText(commentsMock[0].commentText)
                        .parentElement.querySelectorAll("button")[1]
                );
            });

            expect(
                screen.getByRole("button", { name: "Salvar" })
            ).toBeInTheDocument();
            expect(
                screen.getByRole("button", { name: "Cancelar" })
            ).toBeInTheDocument();

            act(() => {
                fireEvent.click(screen.getByRole("button", { name: "Salvar" }));
            });

            await waitFor(() => {
                expect(commentContextValues.editComment).toHaveBeenCalledTimes(
                    1
                );
            });
        });

        it("should call removeComment function on remove comment button click", async () => {
            //change dashboard view option
            act(() => {
                fireEvent.click(screen.getAllByLabelText("View Option")[1]);
            });

            act(() => {
                fireEvent.click(
                    screen
                        .getByText(commentsMock[0].commentText)
                        .parentElement.querySelectorAll("button")[2]
                );
            });

            await waitFor(() => {
                expect(
                    commentContextValues.removeComment
                ).toHaveBeenCalledTimes(1);
            });
        });
    });
});
