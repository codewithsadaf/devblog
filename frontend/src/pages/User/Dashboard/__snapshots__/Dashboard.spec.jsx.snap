// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard/> Rendering should match snapshot 1`] = `
.c7 {
  width: 150px;
  height: 150px;
  object-fit: cover;
  display: block;
  margin: 1em auto;
  border-radius: 7px;
}

.c3 {
  text-transform: none;
  font-weight: bold;
  font-size: 2.9em;
  line-height: 110%;
}

.c10 {
  text-transform: none;
  font-weight: bold;
  font-size: 2.4em;
  line-height: 110%;
}

.c12 {
  text-transform: none;
  font-weight: 500;
  font-size: 1.2em;
  line-height: 120%;
}

.c15 {
  text-transform: none;
  font-weight: bold;
  font-size: 1.2em;
  line-height: 120%;
}

.c25 {
  padding: 0.5em 1.8em;
  font-size: 1em;
  border: none;
  font-weight: bold;
  cursor: pointer;
  background: transparent;
  color: #000;
  border: 2px solid #000;
  border-radius: 4px;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
  padding: 7px 30px;
  border-radius: 4px;
  font-size: 0.8em;
  border: 1px solid;
}

.c25:hover {
  background: #000;
  color: #fff;
  border: 2px solid #000;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c25:hover {
  border: 1px solid;
}

.c0 {
  width: 80%;
  margin: 3em auto 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  margin-bottom: 1.2em;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  -webkit-flex-basis: 40%;
  -ms-flex-preferred-size: 40%;
  flex-basis: 40%;
}

.c5 .c6 {
  width: 300px;
  height: 300px;
  object-fit: fill;
  border-radius: 50%;
}

.c8 {
  -webkit-flex-basis: 58%;
  -ms-flex-preferred-size: 58%;
  flex-basis: 58%;
  min-height: 250px;
  text-align: left;
}

.c8 .c2:nth-of-type(2) {
  margin: 0.3em 0;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c11 {
  font-size: 1.2em;
  margin-left: auto;
  cursor: pointer;
  color: #000;
  opacity: 1;
  -webkit-transition: all 500ms ease-in-out;
  transition: all 500ms ease-in-out;
}

.c11:hover {
  opacity: 0.6;
  -webkit-transition: all 500ms ease-in-out;
  transition: all 500ms ease-in-out;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  border: 1px solid #ccc;
  border-radius: 7px;
  padding: 20px;
  margin-top: 1em;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 2em;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c17 {
  margin: 2em 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 2px solid #000;
}

.c18 {
  background-color: transparent;
  color: #000;
  padding: 7px;
  font-size: 1.2em;
  border-radius: 7px 7px 0 0;
  font-weight: 500;
  cursor: pointer;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
}

.c18.active {
  background-color: #000;
  color: #fff;
  -webkit-transition: all 300ms ease-in-out;
  transition: all 300ms ease-in-out;
  pointer-events: none;
}

.c19 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  border-bottom: 2px solid #ccc;
  font-weight: 700;
  padding: 10px;
}

.c21 {
  font-size: 1em;
  line-height: 150%;
}

.c22 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
  width: 100%;
  border-bottom: 2px solid #eee;
}

.c22 .c20 {
  display: block;
  width: 60%;
  word-wrap: break-word;
}

.c23 {
  margin-left: 2em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23 .c24 {
  margin-left: 0.5em;
}

@media (max-width:768px) {
  .c3 {
    font-size: 1.7em;
    line-height: 120%;
  }
}

@media (max-width:768px) {
  .c0 {
    width: 100%;
  }
}

@media (max-width:768px) {
  .c4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:768px) {
  .c5 .c6 {
    width: 150px;
    height: 150px;
  }
}

@media (max-width:768px) {
  .c5 {
    -webkit-flex-basis: 100%;
    -ms-flex-preferred-size: 100%;
    flex-basis: 100%;
  }
}

@media (max-width:768px) {
  .c8 {
    -webkit-flex-basis: 100%;
    -ms-flex-preferred-size: 100%;
    flex-basis: 100%;
    width: 100%;
    margin-left: 0;
  }
}

@media (max-width:768px) {
  .c22 .c20 {
    width: 100%;
  }
}

@media (max-width:768px) {
  .c22 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

@media (max-width:768px) {
  .c23 .c24 {
    margin-left: 0;
    margin-right: 0.5em;
  }
}

@media (max-width:768px) {
  .c23 {
    margin-left: 0;
    margin: 1em 0;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h1
      class="c2 c3"
    >
      Dashboard
    </h1>
  </div>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <div
        class=""
      >
        <img
          alt="john doe"
          class="c6 c7"
          src="image.png"
        />
      </div>
    </div>
    <div
      class="c8"
    >
      <div
        class="c9"
      >
        <h1
          class="c2 c10"
        >
          john doe
        </h1>
        <a
          class="c11"
          href="/users/edit"
        >
          <svg
            aria-label="Cog Icon"
            class=""
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
            />
          </svg>
        </a>
      </div>
      <h1
        class="c2 c12"
      >
        john@test.com
      </h1>
      <h1
        class="c2 c12"
      >
        
                                    Data de entrada: 5 de Abril 2023
                                
      </h1>
      <div
        class="c13"
      >
        <div
          class="c14"
        >
          <h1
            class="c2 c15"
          >
            Posts
          </h1>
          <h1
            class="c2 c15"
          >
            3
          </h1>
        </div>
        <div
          class="c16"
        >
          <h1
            class="c2 c15"
          >
            Comentários
          </h1>
          <h1
            class="c2 c15"
          >
            2
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-label="View Options"
    class="c17"
  >
    <span
      aria-label="View Option"
      class="c18 active"
    >
      Posts
    </span>
    <span
      aria-label="View Option"
      class="c18"
    >
      Comentários
    </span>
  </div>
  <div
    class="c19"
  >
    <span
      class="c20 c21"
    >
      Título do post
    </span>
    <span
      class="c20 c21"
    >
      Ações
    </span>
  </div>
  <div
    class="c22"
  >
    <span
      class="c20 c21"
    >
      post title 1
    </span>
    <div
      class="c23"
    >
      <button
        class="c24 c25"
        type="button"
      >
        Ler
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Editar
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Excluir
      </button>
    </div>
  </div>
  <div
    class="c22"
  >
    <span
      class="c20 c21"
    >
      post title 2
    </span>
    <div
      class="c23"
    >
      <button
        class="c24 c25"
        type="button"
      >
        Ler
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Editar
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Excluir
      </button>
    </div>
  </div>
  <div
    class="c22"
  >
    <span
      class="c20 c21"
    >
      post title 3
    </span>
    <div
      class="c23"
    >
      <button
        class="c24 c25"
        type="button"
      >
        Ler
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Editar
      </button>
      <button
        class="c24 c25"
        type="button"
      >
        Excluir
      </button>
    </div>
  </div>
</div>
`;
